### PR TITLE
Adds support for restricting channel export to administrator

### PR DESCRIPTION
**Pull request by recontech404**
This PR is part of my Sim project for Mattermost and adds the ability to restrict the `/export` command or api export request to the mattermost system administrators and channel admins.  

You are able to enable/disable the admin user check from the plugin settings in the UI (defaults to disabled).

### CODE REVIEW

**Comment by recontech404** [comment]
There are two existing items that need to be fixed before merging. 

1.  The `model.PermissionManageChannelRoles` (in my testing) does not allow the channel admin (test case being a normal member who created the channel) to export the channel. Only system administrator are able when the feature is enabled.
2. Due to time-constraints, tests still pass successfully but have not been fully updated to reflect the functionality changes (might require a slight refactor to expose certain configuration values).

File: server/plugin.go
Comment (fmartingr) [inline server/plugin.go:87]: Could it be better to have `p.registerAPI` return the handler that goes in there?

**Review by fmartingr — COMMENTED** [review]
I saw you moved `handler.Export` and `handler.hasPermissionToChannel` from `handler` to `plugin`. Any reasoning behind that? Any reason to keep `handler`?

File: server/command_hooks.go
Comment (recontech404) [reply server/command_hooks.go:74]: We can for the API, but in the command hook I am not sure if it makes sense because like you said the user already needs to be part of the channel in order to invoke the command. Also it seems to make more sense sequentially to check if the user even has the correct permissions before we call `p.hasPermissionToChannel` as that has to retrieve the channel in its current implementation, afterwards we then again retrieve the channel in command hook, so that logic seems like it would just introduce a slight performance hit and redundant checks (which will always return true). (https://github.com/mattermost/mattermost-plugin-channel-export/blob/32244afbecfd70d3ff42e3fc61c03a9e93a495b8/server/command_hooks.go#L83)

It may be worth it in terms of simplicity of code but that call is up to the maintainers.

Comment (fmartingr) [reply server/command_hooks.go:74]: Would it makes sense to refactor the new logic into one function that is called from both entrypoints?

Comment (recontech404) [reply server/command_hooks.go:74]: Yeah, I think that makes the most sense given the duplicate functionality.

File: server/api.go
Comment (mgdelacroix) [inline server/api.go:148]: nit: status code returned

```suggestion
			handleError(w, http.StatusForbidden, "user does not have permission", channelID)
```

**Review by mgdelacroix — CHANGES_REQUESTED** [review]
Thanks for the contribution @recontech404!

> No real reason for keeping the handler, I can pull it out but was not sure if there was an exact reason why handler was added in the first place (besides creating a new mutex for the api) so wanted to preserve the original implementation.

I think the reason here is to keep a strong separation of concerns between the plugin itself and the API part of the plugin. I'd propose a different solution: both `router` and `client` parameters of the `registerAPI` method come from the plugin, so instead of making the method part of the plugin, I'd pass the plugin instead of the `router` and `client`, and make `plugin` a property of the `Handler`.

That way, we simplify the signature of the `registerAPI` and the API handler has access to the plugin and its own properties scoped to its own inner workings (in this case the mutexes for example).

We can leave the `Handler.client` property to reduce verbosity or access it through the plugin property to simplify, I'm 1/5 leaning towards removing it.

File: server/plugin.go
Comment (recontech404) [reply server/plugin.go]: Should be updated to have a cleaned up method signature with plugin inside handler.

File: server/command_hooks.go
Comment (recontech404) [reply server/command_hooks.go]: Updated to have a method to remove duplicate code
https://github.com/mattermost/mattermost-plugin-channel-export/blob/11a52c9b4235d56cf5a754cfa49e9e69f0562b13/server/command_hooks.go#L196

File: server/api.go
Comment (recontech404) [reply server/api.go]: Good catch, should be fixed to `http.StatusForbidden`

**Comment by recontech404** [comment]
> Thanks for the contribution @recontech404!
> 
> > No real reason for keeping the handler, I can pull it out but was not sure if there was an exact reason why handler was added in the first place (besides creating a new mutex for the api) so wanted to preserve the original implementation.
> 
> I think the reason here is to keep a strong separation of concerns between the plugin itself and the API part of the plugin. I'd propose a different solution: both `router` and `client` parameters of the `registerAPI` method come from the plugin, so instead of making the method part of the plugin, I'd pass the plugin instead of the `router` and `client`, and make `plugin` a property of the `Handler`.
> 
> That way, we simplify the signature of the `registerAPI` and the API handler has access to the plugin and its own properties scoped to its own inner workings (in this case the mutexes for example).
> 
> We can leave the `Handler.client` property to reduce verbosity or access it through the plugin property to simplify, I'm 1/5 leaning towards removing it.

Thanks for the suggestions and I agree on the making `plugin` part of the `Handler`. Makes things a bit easier to read. The code should be updated to reflect those changes.

File: server/command_hooks.go
Comment (fmartingr) [inline server/command_hooks.go:199]: You need to check the `model.permissionManageChannelRoles` for the channel scope, not system scope: https://developers.mattermost.com/integrate/reference/server/server-reference/#API.HasPermissionToChannel

That should make that check work.

**Review by fmartingr — CHANGES_REQUESTED** [review]
Found out the issue with the permission check.

File: server/command_hooks.go
Comment (recontech404) [reply server/command_hooks.go]: Great thanks! I tested that and it fixes the issue so the code should be updated.

**Review by fmartingr — APPROVED** [review]
Looking amazing. We managed to reduce complexity and narrow it down and make it work. Thank you!

**Review by mgdelacroix — APPROVED** [review]
LGTM, thanks for all your work on this @recontech404!!

**Comment by wiggin77** [comment]
@AayushChaudhary0001 this got merged without QA.  Can you please QA this new feature as we are going to bundle the plugin with the next server release.

**Comment by AayushChaudhary0001** [comment]
@wiggin77 I have tested this PR and this had an issue that the default setting was set to false according to the code, but on uploading the plugin first time on a server, the default setting in system console to restrict to system admin and channel admin was always true which was not matching, hence this was not working fine.

**Pull request merged**
